infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/msgHandlers/ReplicationHandlerMethods.java

```diff
@@ -0,0 +1,192 @@
+     * be fast and not block. If a handler blocks for an extended period of time, it will
+     * exhaust the server's thread pool. I/O and other long operations should be handled
+     * on another thread.
+     */
+    @FunctionalInterface
+    public interface HandlerMethod {
+        void handle(CorfuMessage.RequestMsg req,
+                    CorfuMessage.ResponseMsg res,
+                    @Nonnull IClientServerRouter r);
+    }
+
+    /**
+     * Construct a new instance of ReplicationHandlerMethods.
+     */
+    public ReplicationHandlerMethods() {
+        requestHandlerMap = new ConcurrentHashMap<>();
+        responseHandlerMap = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Given a {@link CorfuMessage.RequestMsg} return the corresponding handler.
+     *
+     * @param request   The request message to handle.
+     * @return          The appropriate {@link HandlerMethod}.
+     */
+    protected HandlerMethod getRequestHandler(CorfuMessage.RequestMsg request) {
+        return requestHandlerMap.get(request.getPayload().getPayloadCase());
+    }
+
+    /**
+     * Given a {@link CorfuMessage.ResponseMsg} return the corresponding handler.
+     *
+     * @param response   The respose message to handle.
+     * @return          The appropriate {@link HandlerMethod}.
+     */
+    protected HandlerMethod getResponseHandler(CorfuMessage.ResponseMsg response) {
+        return responseHandlerMap.get(response.getPayload().getPayloadCase());
+    }
+
+    /**
+     * Handle an incoming Corfu request message.
+     *
+     * @param req       The request message to handle.
+     * @param r         The server router.
+     */
+    @SuppressWarnings("unchecked")
+    public void handle(CorfuMessage.RequestMsg req, CorfuMessage.ResponseMsg res, IClientServerRouter r) {
+        final HandlerMethod handler = req != null ? getRequestHandler(req) : getResponseHandler(res);
+
+        try {
+            handler.handle(req, res, r);
+        } catch (Exception e) {
+            if (req != null) {
+                log.error("handle[{}]: Unhandled exception processing {} request",
+                        req.getHeader().getRequestId(), req.getPayload().getPayloadCase(), e);
+            } else {
+                log.error("handle[{}]: Unhandled exception processing {} response",
+                        res.getHeader().getRequestId(), res.getPayload().getPayloadCase(), e);
+            }
+        }
+    }
+
+    /**
+     * Generate handlers for a particular server.
+     *
+     * @param caller    The context that is being used. Call MethodHandles.lookup() to obtain.
+     * @param server    The object that implements the server.
+     * @return          New request handlers for caller class.
+     */
+    public static ReplicationHandlerMethods generateHandler(@Nonnull final MethodHandles.Lookup caller,
+                                                            @NonNull final LogReplicationAbstractServer server) {
```

Comment:
Why this method has to be static? I thought the logic could be moved to constructor right? and also the method L88 of `LogReplicationServer` is redundant IMO. Maybe we can cleanup these and do something like:
```
private final ReplicationHandlerMethods handlerMethods = 
    new ReplicationHandlerMethods().generateHandler();
```
in LogReplicationServer